pkg/proc/eval.go

```diff
@@ -3005,6 +3001,148 @@ func (v *Variable) reslice(low int64, high int64, trustLen bool) (*Variable, err
+				// Not a closure, normal function
+				if _, ok := v.bi.PackageMap[vname]; ok {
+					return nil, fmt.Errorf("package %s has no function %s", vname, name)
+				}
+				return nil, fmt.Errorf("%s has no member %s", vname, name)
+			}
+			return nil, v.Unreadable
+		}
+		if v.closureAddr != 0 {
+			fn = v.bi.PCToFunc(v.Base)
+			if fn != nil {
+				cst := fn.extra(v.bi).closureStructType
+				v = v.newVariable(v.Name, v.closureAddr, cst, v.mem)
+				closure = true
+			}
+		}
+	}
+
+	queue := []*Variable{v}
+	seen := map[string]struct{}{}
+	first := true
+
+	for len(queue) > 0 {
+		v := queue[0]
+		queue = append(queue[:0], queue[1:]...)
+		if _, isseen := seen[v.RealType.String()]; isseen {
+			continue
+		}
+		seen[v.RealType.String()] = struct{}{}
+
+		typ := v.DwarfType
+		ptyp, isptr := typ.(*godwarf.PtrType)
+		if isptr {
+			typ = ptyp.Type
+		}
+
+		var pkg, receiver string
+		typePath := typ.Common().Name
+		dot := strings.LastIndex(typePath, ".")
+		if dot >= 0 {
+			pkg, receiver = typePath[:dot], typePath[dot+1:]
+		}
+
+		if len(pkg) > 0 && len(receiver) > 0 {
+			rv, err := lookupMethod(v, isptr, pkg, receiver, name)
+			if err != nil {
+				return nil, err
+			} else if rv != nil {
+				return rv, nil
+			}
+		}
+
+		// queue embedded fields for search
+		structVar := v.maybeDereference()
+		structVar.Name = v.Name
+		if structVar.Unreadable != nil {
+			return structVar, nil
+		}
+		switch t := structVar.RealType.(type) {
+		case *godwarf.StructType:
+			for _, field := range t.Field {
+				if field.Name == name {
+					return structVar.toField(field)
+				}
+				if len(queue) == 0 && field.Name == "&"+name && closure {
+					f, err := structVar.toField(field)
+					if err != nil {
+						return nil, err
+					}
+					return f.maybeDereference(), nil
+				}
+				isEmbeddedStructMember :=
+					field.Embedded ||
+						(field.Type.Common().Name == field.Name) ||
+						(len(field.Name) > 1 &&
+							field.Name[0] == '*' &&
+							field.Type.Common().Name[1:] == field.Name[1:])
+				if !isEmbeddedStructMember {
+					continue
+				}
+				embeddedVar, err := structVar.toField(field)
+				if err != nil {
+					return nil, err
+				}
+				// Check for embedded field referenced by type name
+				parts := strings.Split(field.Name, ".")
+				if len(parts) > 1 && parts[1] == name {
+					return embeddedVar, nil
+				}
+				embeddedVar.Name = structVar.Name
+				queue = append(queue, embeddedVar)
+			}
+		case *godwarf.InterfaceType:
+			v.loadInterface(0, false, LoadConfig{})
+			if len(v.Children) > 0 {
+				if rv, _ := v.Children[0].findMethod(name); rv != nil {
+					return rv, nil
+				}
+			}
+		default:
+			if first {
+				return nil, fmt.Errorf("%s (type %s) is not a struct", vname, structVar.TypeString())
+			}
+		}
+		first = false
+	}
+
+	return nil, fmt.Errorf("%s has no member %s", vname, name)
+}
+
 // findMethod finds method mname in the type of variable v
 func (v *Variable) findMethod(mname string) (*Variable, error) {
```

Comment:
This now never gets called except through the code of findStructMemberOrMethod, it should be deleted (it's a lot of duplicated code) and a flag should be added to findStructMemberOrMethod that enables or disables struct member lookup.